gcp/modules/gcp-stackdriver-lbm/resources/log_based_metrics/compute_instances_insert.json

```diff
@@ -0,0 +1,4 @@
+{
+  "name": "compute.instances.insert",
+  "filter": "resource.type=\"gce_instance\" AND jsonPayload.event_subtype=\"compute.instances.insert\""
```

Comment:
I double checked – `protoPayload.methodName="v1.compute.instances.insert"` is not catching instances creation events from GCP console. Also, I am able to see TF-managed instances creation events with current filter [in my dev project](https://console.cloud.google.com/logs/viewer?authuser=1&organizationId=327626828918&project=gpii-gcp-dev-natarajaya&supportedpurview=project&minLogLevel=0&expandAll=false&timestamp=2019-05-22T21:10:26.734000000Z&customFacets=&limitCustomFacetWidth=true&dateRangeStart=2019-05-22T15:10:26.985Z&dateRangeEnd=2019-05-22T21:10:26.985Z&interval=PT6H&resource=gce_instance&scrollTimestamp=2019-05-22T18:16:44.944926000Z&advancedFilter=resource.type%3D%22gce_instance%22%20AND%20jsonPayload.event_subtype%3D%22compute.instances.insert%22%0A).

Reply:
Pls. check on PRD:
 - [the `jsonPayload` filter does not catch anything](https://console.cloud.google.com/logs/viewer?project=gpii-gcp-prd&authuser=2&organizationId=247149361674&supportedpurview=project&minLogLevel=0&expandAll=false&timestamp=2019-05-23T12:14:15.323000000Z&interval=NO_LIMIT&scrollTimestamp=2019-05-23T11:07:12.155150783Z&advancedFilter=resource.type%3D%22gce_instance%22%20AND%20jsonPayload.event_subtype%3D%22compute.instances.insert%22)
- [the `protoPayload` one does catch a lot](https://console.cloud.google.com/logs/viewer?project=gpii-gcp-prd&authuser=2&organizationId=247149361674&supportedpurview=project&minLogLevel=0&expandAll=false&timestamp=2019-05-23T12:04:22.551000000Z&interval=NO_LIMIT&scrollTimestamp=2019-03-25T03:13:38.729000000Z&dateRangeStart=2019-01-01T13:04:06.000Z&dateRangeEnd=2019-05-23T12:04:06.343Z&advancedFilter=resource.type%3D%22gce_instance%22%20AND%20protoPayload.methodName%3D%22v1.compute.instances.insert%22).

I addition:
- you're correct that instance created via web UI is only caught with the `jsonPayload` filter, it also shows up with `protoPayload` if the method name is `protoPayload.methodName="beta.compute.instances.insert"`, so filter like `protoPayload.methodName:"compute.instances.insert"` would catch all
- instance created via `gcloud` command is caught by both
- instances created though instance group are catched by both


After looking at the logs more in detail, the `jsonPayload` logs come from compute api logs (`projects/gpii-gcp-dev-stepan/logs/compute.googleapis.com%2Factivity_log`, the `protoPayload` logs come from  audit logs (`type.googleapis.com/google.cloud.audit.AuditLog`). So I guess the difference in prd is caused by retention period (30d for normal vs 400d for audit logs).

I would say that the audit logs due to their nature - mainly the fact that they can't be disabled on individual projects - are better for this purpose. Otherwise, I can easily go to [Logs settings](https://console.cloud.google.com/logs/usage?authuser=2&project=gpii-gcp-dev-stepan) and disable the logs before creating the VM.

Reply:
Cool! Thanks for this research! The fact that audit logs can't be disabled on project level is a great argument in favor of using audit logs. I'll also update the filter to include beta API in condition.